### PR TITLE
Update the HIL client interface to use HIL's networking action APIs t…

### PR DESCRIPTION
**Pull request by naved001**
…o query network calls.

* Also handles some more edge cases.
* End to end tests for the hil client interface passed.

Would appreciate your reviews @SahilTikale @Izhmash

### CODE REVIEW

File: common/ulsr_hil_client.py [hunk 1]
```diff
@@ -29,10 +29,12 @@ class ProjectMismatchError(Exception):
...
 
-def _hil_client_connect(endpoint_ip, name, pw):
-    '''
-    Connect to the HIL server and return a HIL Client instance
-    Note this call will succeed if the API server is running, but the network server is down           '''
+def hil_client_connect(endpoint_ip, name, pw):
+    """"
+    Returns an instance of HIL client.
+
+    This call does now check (or connect to) the HIL server.
```
Comment (ianballou) [inline common/ulsr_hil_client.py]: What did you mean by "does now check"?

**Review by ianballou — COMMENTED** [review]
This looks good so far, I'm going to review again in the morning with fresh eyes.

File: common/ulsr_hil_client.py (hunk 1 above)
Comment (naved001) [reply common/ulsr_hil_client.py]: my bad, I meant "does not check". I will update this.

Comment (naved001) [reply common/ulsr_hil_client.py]: fixed

File: common/ulsr_hil_client.py [hunk 2]
```diff
@@ -309,8 +323,51 @@ def connect_node_to_project(hil_client, node, project):
...
+    Returns a status_id to poll on
+    """
+    try:
+        response = hil_client.node.connect_network(
+            node, nic, network, 'vlan/native')
+        status_id = response['status_id']
+        log_info('Node `%s` succesfully connected to network `%s`' %
+                 (node, network))
+        return status_id
+    except FailedAPICallException as e:
```
Comment (ianballou) [inline common/ulsr_hil_client.py:342]: For these functions that have the two `except`s at the end, was it your plan for unexpected errors to just show their output?

Comment (naved001) [reply common/ulsr_hil_client.py:342]: Yes, I am not catching unexpected errors because we don't know how to handle the unknown. If/when an unexpected error is raised, that will just stop the code and the message will be loud and clear (will be useful for debugging).

File: common/ulsr_hil_client.py [hunk 3]
```diff
@@ -29,10 +29,12 @@ class ProjectMismatchError(Exception):
...
 
-def _hil_client_connect(endpoint_ip, name, pw):
-    '''
-    Connect to the HIL server and return a HIL Client instance
-    Note this call will succeed if the API server is running, but the network server is down           '''
+def hil_client_connect(endpoint_ip, name, pw):
+    """"
+    Returns an instance of HIL client.
+
+    This call does not check (or connect to) the HIL server.
```
Comment (ianballou) [inline common/ulsr_hil_client.py]: Maybe we could add a little more comment context here, it's a little weird to have `hil_client_connect` and then say it doesn't connect to HIL server.

File: common/ulsr_hil_client.py [hunk 4]
```diff
@@ -309,8 +323,51 @@ def connect_node_to_project(hil_client, node, project):
...
-        log_error('Unable to connect node `%s` to project `%s`' % (node, project))
+        log_error('Unable to connect node `%s` to project `%s`' %
+                  (node, project))
         raise HILClientFailure(e)
     except ConnectionError as e:
         log_error("connect_node_to_project: Couldn't connect to HIL server.")
         raise HILClientFailure(e)
+
+
+def connect_network(hil_client, node, nic, network, channel):
```
Comment (ianballou) [inline common/ulsr_hil_client.py:329]: should we use the underscore notation to mark that it's a helper?

File: common/ulsr_hil_client.py (hunk 2 above)
Comment (Izhmash) [reply common/ulsr_hil_client.py]: Sounds good

**Review by ianballou — COMMENTED** [review]
I still can't find anything wrong, just little picky things here and there.  How did you test it?

**Review by ianballou — APPROVED** [review]

File: common/ulsr_hil_client.py (hunk 4 above)
Comment (naved001) [reply common/ulsr_hil_client.py:329]: From the [docs](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) 
> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose name starts with an underscore.

This helper can actually be imported in some other module, so I don't really need to actively stop it from being imported.

File: common/ulsr_hil_client.py (hunk 3 above)
Comment (naved001) [reply common/ulsr_hil_client.py]: good point, I'll change the name of the method to make it clear.

**Comment by naved001** [comment]
> How did you test this?

https://github.com/mghpcc-projects/user_level_slurm_reservations/blob/master/test/hil_client_test.py

This module calls the reserve and release methods and tests it against a real HIL server. There aren't many negative tests, because these methods correct the inconsistencies and move forward with it.

File: common/ulsr_hil_client.py (hunk 3 above)
Comment (naved001) [reply common/ulsr_hil_client.py]: okay, fixed this. I now call it `hil_client_object`

File: common/ulsr_hil_client.py [hunk 5]
```diff
@@ -29,10 +29,8 @@ class ProjectMismatchError(Exception):
     """Raised when projects don't match"""
 
 
-def _hil_client_connect(endpoint_ip, name, pw):
-    '''
-    Connect to the HIL server and return a HIL Client instance
-    Note this call will succeed if the API server is running, but the network server is down           '''
+def hil_client_object(endpoint_ip, name, pw):
+    """"Returns a HIL client object"""
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py]: doc string has 4 " should be 3 "

File: common/ulsr_hil_client.py [hunk 6]
```diff
@@ -50,7 +48,7 @@ def hil_init():
     '''
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py]: Why is this empty three colons here? any specific reason.

File: common/ulsr_hil_client.py [hunk 7]
```diff
@@ -13,7 +13,7 @@
 from hil.client.base import FailedAPICallException
 from ulsr_logging import log_info, log_debug, log_error
 from ulsr_settings import HIL_ENDPOINT, HIL_USER, HIL_PW, OBM_NIC, \
- OBM_NETWORK, MAINTENANCE_PROJECT
+    OBM_NETWORK, MAINTENANCE_PROJECT
 from ulsr_helpers import is_hil_available
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py:17]: Where is this `ulsr_helpers` function located? I need to see the function `is_hil_available`

File: common/ulsr_hil_client.py [hunk 8]
```diff
@@ -50,7 +48,7 @@ def hil_init():
     '''
     '''
     if is_hil_available():
-        status =_hil_client_connect(HIL_ENDPOINT, HIL_USER, HIL_PW)
+        status = hil_client_object(HIL_ENDPOINT, HIL_USER, HIL_PW)
     else:
         log_info('HIL unavailable, all HIL operations will appear to succeed')
         status = True
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py:54]: Assuming that the function is_hil_available returns true, what does `hil_client_object` return into the variable `status`. Does it return `false` ?

why is the status in the `else` branch set to TRUE?

File: common/ulsr_hil_client.py [hunk 9]
```diff
@@ -94,35 +92,42 @@ def hil_reserve_nodes(nodelist, from_project, hil_client=None):
...
         elif project != from_project and project != MAINTENANCE_PROJECT:
-            log_error('HIL reservation failure: Node `%s` (in project `%s`) not in `%s` project' % (node, project, from_project))
+            log_error('HIL reservation failure: Node `%s` (in project `%s`) not in `%s` project' % (
+                node, project, from_project))
             raise ProjectMismatchError()
 
     # Power off all nodes.
     # Check if the obm network is connected, if not, connect it.
     for node in nodelist:
         if not _is_network_connected(hil_client, node, OBM_NETWORK, 'vlan/native'):
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py:106]: Why is HIL connecting to the OBM network? Isn't that a design violation. I thought the OBM network was never exposed to the users of HIL.  Please explain what is happening here? Can any unprivileged user execute this code (given here)?

File: common/ulsr_hil_client.py [hunk 10]
```diff
@@ -94,35 +92,42 @@ def hil_reserve_nodes(nodelist, from_project, hil_client=None):
...
+    # Finally, remove node from `from_project` and MAINTENANCE_PROJECT.
+    for node in nodelist:
         node_info = show_node(hil_client, node)
         project = node_info['project']
 
+        if project == from_project:
+            remove_node_from_project(hil_client, node, from_project)
+
         if project == MAINTENANCE_PROJECT:
             remove_node_from_project(hil_client, node, MAINTENANCE_PROJECT)
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py]: Suggestion:
what if you write it like 
```
if project in [from_project, MAINTENANCE_PROJECT]:
    remove_node_from_project(hil_client, node, project)
```

File: common/ulsr_hil_client.py [hunk 11]
```diff
@@ -94,35 +92,42 @@ def hil_reserve_nodes(nodelist, from_project, hil_client=None):
...
-        remove_node_from_project(hil_client, node, from_project)
+    # Check the status of the networking actions after we have queued all
+    # operations in the previous step.
+    for status_id in status_ids:
+        assert_network_operation_success(hil_client, status_id)
 
+    # Finally, remove node from `from_project` and MAINTENANCE_PROJECT.
+    for node in nodelist:
         node_info = show_node(hil_client, node)
         project = node_info['project']
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py:127]: Why are there two `if` clauses? Can a node be part of two projects that you have to do independent checks for them?

File: common/ulsr_hil_client.py [hunk 12]
```diff
@@ -165,91 +170,95 @@ def hil_free_nodes(nodelist, to_project, hil_client=None):
...
 
         if (project == to_project):
-            log_info('HIL release: Node `%s` already in `%s` project, skipping' % (node, to_project))
+            log_info('HIL release: Node `%s` already in `%s` project, skipping' % (
+                node, to_project))
             nodelist.remove(node)
         elif (project is not None and project != MAINTENANCE_PROJECT):
             # if a node was not in the free pool or the maintenance project.
             # we remove it from that project, which puts it in the maintenance pool.
```
Comment (SahilTikale) [inline common/ulsr_hil_client.py:177]: Does removing a node from a project automatically moves it to the maintenance pool, rather than a free pool?
I am assuming you wanted to say `we remove ''and'' put` instead of `we remove which puts`

**Comment by SahilTikale** [comment]
I have made a first pass on the PR. Asked few 101 questions. You can respond to them or we can talk about it when we meet.

**Review by SahilTikale — CHANGES_REQUESTED** [review]

File: common/ulsr_hil_client.py (hunk 9 above)
Comment (naved001) [reply common/ulsr_hil_client.py:106]: Not really a design violation. HIL users can choose to control whatever networks and switches they choose. @christophernhill preferred controlling the OBM networks with HIL. This way nodes in the free pool are disconnected from the OBM networks too (for added security).

Also, this is privileged code that is only run by the ulsr (user level slurm reservation) code. No regular users can execute this.

File: common/ulsr_hil_client.py (hunk 10 above)
Comment (naved001) [reply common/ulsr_hil_client.py]: yeah, that sounds better. I'll do that.

File: common/ulsr_hil_client.py (hunk 11 above)
Comment (naved001) [reply common/ulsr_hil_client.py:127]: Just in case something else goes south (manual intervention by HIL admins for instance), I am checking all the possibilities; so yes it could either be a part of the maintenance project or the given `from_project`

File: common/ulsr_hil_client.py (hunk 12 above)
Comment (naved001) [reply common/ulsr_hil_client.py:177]: > Does removing a node from a project automatically moves it to the maintenance pool, rather than a free pool?

Yes, it does so automatically. This feature was added by @Izhmash. So my comment "we remove which puts" is accurate.

File: common/ulsr_hil_client.py (hunk 8 above)
Comment (naved001) [reply common/ulsr_hil_client.py:54]: Regardless of the value of `is_hil_available()`, _hil_client_object will just return a HIL HTTP client object (just a regular python object). 

I didn't understand the original intent of using `status` to store the object and also returning the status, which is why I didn't change it. @tpd001 had some plans on such error handling.

File: common/ulsr_hil_client.py (hunk 7 above)
Comment (naved001) [reply common/ulsr_hil_client.py:17]: [`is_hil_available`](https://github.com/mghpcc-projects/user_level_slurm_reservations/blob/development/common/ulsr_helpers.py#L414y).

File: common/ulsr_hil_client.py (hunk 10 above)
Comment (naved001) [reply common/ulsr_hil_client.py]: This is actually right. Now I remember why I did that. Because if the node was in a project, then calling remove will put it in the maintenance project and then you have to call remove again to finally put it in the free pool. And in case the node was already in the maintenance project, it skips the first if clause and goes to the second one. Does that make sense?

**Comment by naved001** [comment]
@SahilTikale Updated with changes you suggested.

File: common/ulsr_hil_client.py (hunk 6 above)
Comment (SahilTikale) [reply common/ulsr_hil_client.py]: If it is a docstring, please fix it.

File: common/ulsr_hil_client.py (hunk 8 above)
Comment (SahilTikale) [reply common/ulsr_hil_client.py:54]: I understand @tpd001  wanted it that way, the motivation for coding this way may not be clear to you.
Can you make it more pythonic?

File: common/ulsr_hil_client.py (hunk 9 above)
Comment (SahilTikale) [reply common/ulsr_hil_client.py:106]: ok

File: common/ulsr_hil_client.py (hunk 4 above)
Comment (SahilTikale) [reply common/ulsr_hil_client.py:329]: if you think that this can be used in some other module, wouldn't it be a good idea to move it to a file with such common definitions?
I agree with @Izhmash that if it is a private method add an '_' to it
or if you think it can be used by other modules, move all such functions to a common library file.

**Review by SahilTikale — CHANGES_REQUESTED** [review]

**Comment by SahilTikale** [comment]
Just checking, is this work updated/done ?

**Comment by naved001** [comment]
This does work, but I haven't had the time to make changes that you
suggested.
